internal/pkg/skuba/addons/dex.go

```diff
@@ -45,31 +46,44 @@ func renderDexTemplate(addonConfiguration AddonConfiguration) string {
 
 type dexCallbacks struct{}
 
-func (dexCallbacks) beforeApply(addonConfiguration AddonConfiguration, skubaConfiguration *skuba.SkubaConfiguration) error {
-	var err error
-
-	client, err := kubernetes.GetAdminClientSet()
+func (dexCallbacks) beforeApply(client clientset.Interface, addonConfiguration AddonConfiguration, skubaConfiguration *skuba.SkubaConfiguration) error {
+	// handles oidc client-secret
+	exist, err := oidc.IsSecretExist(client, oidc.ClientSecretName)
 	if err != nil {
-		return errors.Wrap(err, "could not get admin client set")
+		return errors.Wrap(err, "unable to determine if oidc client-secret exists")
+	}
+	if !exist {
+		// generate client secret with length=12
+		// client secret is used by auth client (gangway) to authenticate to auth server (dex)
+		clientSecret, err := oidc.RandomGenerateWithLength(12)
+		if err != nil {
+			return errors.Wrap(err, "unable to generate oidc client-secret")
+		}
+		err = oidc.CreateOrUpdateToSecret(client, oidc.ClientSecretName, oidc.ClientSecretKey_Gangway, clientSecret)
+		if err != nil {
+			return err
+		}
 	}
 
-	dexCertExists, err := dex.DexCertExists(client)
+	// handles dex certificates
+	// TODO: should we keep rotates server certificate when addon upgrade?
+	exist, err = oidc.IsSecretExist(client, oidc.DexCertSecretName)
 	if err != nil {
 		return errors.Wrap(err, "unable to determine if dex certificate exists")
 	}
-	err = dex.CreateCert(client, skubaconstants.PkiDir(), skubaconstants.KubeadmInitConfFile())
-	if err != nil {
-		return errors.Wrap(err, "unable to create dex certificate")
+	if err = oidc.CreateServerCert(client, skubaconstants.PkiDir(), oidc.DexCertCN, addonConfiguration.ControlPlaneHost, oidc.DexCertSecretName); err != nil {
```

Comment:
Maybe there is no need to deal with the cert rotation anymore :), just leave to our new solution `addon cert rotation` to have a single way to do that and this will also avoid any confusing. 

Also, if users choose to change the cert of `dex` or `gangway`, this will possibly overrite the certs with the wrong CA. 

The same suggestion to `gangway.`

Reply:
Agree, I forgot custom CA case, I'll remove it, thx.